### PR TITLE
[SPARK-31134][SQL] optimize skew join after shuffle partitions are coalesced

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

Run the `OptimizeSkewedJoin` rule after the `CoalesceShufflePartitions` rule.

### Why are the changes needed?

Remove duplicated coalescing code in `OptimizeSkewedJoin`.

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @maryannxue @JkSelf

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 1]
```diff
@@ -328,6 +287,48 @@ case class OptimizeSkewedJoin(conf: SQLConf) extends Rule[SparkPlan] {
...
+  }
+
+  private def getMapStats(stage: ShuffleQueryStageExec): MapOutputStatistics = {
+    assert(stage.resultOption.isDefined, "ShuffleQueryStageExec should" +
+      " already be ready when executing OptimizeSkewedPartitions rule")
+    stage.resultOption.get.asInstanceOf[MapOutputStatistics]
+  }
+}
+
+private case class ShuffleStageInfo(
```
Comment (maryannxue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala]: nit: partitions -> partitionsWithSizes?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 2]
```diff
@@ -328,6 +287,48 @@ case class OptimizeSkewedJoin(conf: SQLConf) extends Rule[SparkPlan] {
...
+        case (size, i) => CoalescedPartitionSpec(i, i + 1) -> size
+      }
+      Some(ShuffleStageInfo(s, mapStats, partitions))
+
+    case CustomShuffleReaderExec(s: ShuffleQueryStageExec, partitionSpecs, _) =>
+      val mapStats = getMapStats(s)
+      val sizes = mapStats.bytesByPartitionId
+      val partitions = partitionSpecs.map {
+        case spec @ CoalescedPartitionSpec(start, end) =>
+          var sum = 0L
```
Comment (maryannxue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala]: nit: sizes.slice(start, end).sum?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 3]
```diff
@@ -150,72 +146,65 @@ case class OptimizeSkewedJoin(conf: SQLConf) extends Rule[SparkPlan] {
...
-              rightSidePartitions += p
-            }
-            nonSkewPartitionIndices.clear()
-          }
-
           val leftParts = if (isLeftSkew) {
-            val mapStartIndices = getMapStartIndices(left, partitionIndex, leftTargetSize)
-            if (mapStartIndices.length > 1) {
+            val CoalescedPartitionSpec(start, end) = left.partitions(partitionIndex)._1
+            assert(start + 1 == end, "coalesced partition should never be skewed.")
```
Comment (maryannxue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala]: First of all, our factor check is not strict enough as being "> 0", what happens here if it's set to "1"?
Second, the assert is usually disabled in production, which could lead to errors later in this code.
We should probably make it more robust by putting this condition into `isSkew`. And you can still add such an assertion in `isSkew` implementation.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 4]
```diff
@@ -150,72 +146,65 @@ case class OptimizeSkewedJoin(conf: SQLConf) extends Rule[SparkPlan] {
...
       val leftSkewDesc = new SkewDesc
       val rightSkewDesc = new SkewDesc
       for (partitionIndex <- 0 until numPartitions) {
-        val leftSize = leftStats.bytesByPartitionId(partitionIndex)
+        val leftSize = leftSizes(partitionIndex)
         val isLeftSkew = isSkewed(leftSize, leftMedSize) && canSplitLeft
-        val rightSize = rightStats.bytesByPartitionId(partitionIndex)
+        val rightSize = rightSizes(partitionIndex)
         val isRightSkew = isSkewed(rightSize, rightMedSize) && canSplitRight
         if (isLeftSkew || isRightSkew) {
```
Comment (maryannxue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala]: Not related to this PR, but I think we can remove this outer `if` now.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala]: `slice` will create a new array, which is less efficient.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 5]
```diff
@@ -150,72 +146,65 @@ case class OptimizeSkewedJoin(conf: SQLConf) extends Rule[SparkPlan] {
...
+            specsAfterSplit.getOrElse(Seq(left.partitions(partitionIndex)._1))
           } else {
-            Seq(CoalescedPartitionSpec(partitionIndex, partitionIndex + 1))
+            Seq(left.partitions(partitionIndex)._1)
           }
 
           val rightParts = if (isRightSkew) {
-            val mapStartIndices = getMapStartIndices(right, partitionIndex, rightTargetSize)
-            if (mapStartIndices.length > 1) {
+            val CoalescedPartitionSpec(start, end) = right.partitions(partitionIndex)._1
```
Comment (JkSelf) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala]: The code in the calculation of `leftParts ` and `rightParts ` is almost same. It is better to wrap the code in a method.

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 6]
```diff
@@ -150,146 +154,93 @@ case class OptimizeSkewedJoin(conf: SQLConf) extends Rule[SparkPlan] {
...
+        val isLeftSkew = isSkewed(leftActualSizes(partitionIndex), leftMedSize) && canSplitLeft
+        val leftPartSpec = left.partitionsWithSizes(partitionIndex)._1
+        val isLeftCoalesced = leftPartSpec.startReducerIndex + 1 < leftPartSpec.endReducerIndex
+
+        val isRightSkew = isSkewed(rightActualSizes(partitionIndex), rightMedSize) && canSplitRight
+        val rightPartSpec = right.partitionsWithSizes(partitionIndex)._1
+        val isRightCoalesced = rightPartSpec.startReducerIndex + 1 < rightPartSpec.endReducerIndex
+
+        // Ideally a skewed partition won't get coalesced, but skip it here for safety.
+        val leftParts = if (isLeftSkew && !isLeftCoalesced) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala]: @JkSelf I tried to create a common method to handle both sides, but the method takes too many parameters so I give up. Besides, it's not much duplicated code here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 7]
```diff
@@ -150,146 +154,93 @@ case class OptimizeSkewedJoin(conf: SQLConf) extends Rule[SparkPlan] {
...
-      val numPartitions = leftStats.bytesByPartitionId.length
-
-      val leftMedSize = medianSize(leftStats)
-      val rightMedSize = medianSize(rightStats)
+      assert(left.partitionsWithSizes.length == right.partitionsWithSizes.length)
+      val numPartitions = left.partitionsWithSizes.length
+      // We use the median size of the original shuffle partitions to detect skewed partitions.
+      val leftMedSize = medianSize(left.mapStats)
+      val rightMedSize = medianSize(right.mapStats)
       logDebug(
```
Comment (maryannxue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala]: Again, unrelated to this PR: this log is verbose (literally every partition will be printed, that can be a long list) yet not really informative (what's the result of skew join handling to that partition??).
We can either move it after each group of skew partitions is created and log skewed specs, or remove it at all.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 8]
```diff
@@ -150,146 +154,93 @@ case class OptimizeSkewedJoin(conf: SQLConf) extends Rule[SparkPlan] {
...
-            Seq(CoalescedPartitionSpec(partitionIndex, partitionIndex + 1))
+        val isLeftSkew = isSkewed(leftActualSizes(partitionIndex), leftMedSize) && canSplitLeft
+        val leftPartSpec = left.partitionsWithSizes(partitionIndex)._1
+        val isLeftCoalesced = leftPartSpec.startReducerIndex + 1 < leftPartSpec.endReducerIndex
+
+        val isRightSkew = isSkewed(rightActualSizes(partitionIndex), rightMedSize) && canSplitRight
+        val rightPartSpec = right.partitionsWithSizes(partitionIndex)._1
+        val isRightCoalesced = rightPartSpec.startReducerIndex + 1 < rightPartSpec.endReducerIndex
+
+        // Ideally a skewed partition won't get coalesced, but skip it here for safety.
```
Comment (maryannxue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala]: nit: A skewed partition should never be coalesced, but skip it here just to be safe.

**Review by maryannxue — APPROVED** [review]
LGTM.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 9]
```diff
@@ -150,146 +155,97 @@ case class OptimizeSkewedJoin(conf: SQLConf) extends Rule[SparkPlan] {
...
-            Seq(CoalescedPartitionSpec(partitionIndex, partitionIndex + 1))
+        val isLeftSkew = isSkewed(leftActualSizes(partitionIndex), leftMedSize) && canSplitLeft
+        val leftPartSpec = left.partitionsWithSizes(partitionIndex)._1
+        val isLeftCoalesced = leftPartSpec.startReducerIndex + 1 < leftPartSpec.endReducerIndex
+
+        val isRightSkew = isSkewed(rightActualSizes(partitionIndex), rightMedSize) && canSplitRight
+        val rightPartSpec = right.partitionsWithSizes(partitionIndex)._1
+        val isRightCoalesced = rightPartSpec.startReducerIndex + 1 < rightPartSpec.endReducerIndex
+
+        // A skewed partition should never be coalesced, but skip it here just to be safe.
```
Comment (Ngone51) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala]: Say we have original map output: 100, 10, 2000, and the coalesce target is 100. So, after `CoalesceShufflePartitions`, we shall have `CoalescedPartitionSpec(0, 1)` and `CoalescedPartitionSpec(1, 3)`. Then, we start to apply `OptimizeSkewedJoin` where `CoalescedPartitionSpec(1, 3)` is obviously skewed but can be missed. Right?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala]: I don't think the coalesce rule will coalesce `10` and `2000`, can you double check?

Comment (Ngone51) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala]: Oh yeah, I checked, you're right!

**Comment by Ngone51** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merged to master/3.0

**Pull request closed without merging**